### PR TITLE
Yul type checking

### LINKED ISSUE #7070: [Yul] add types to dialects, but keep a default type
Since eWasm has different types, we have to enable the type system in Yul before eWasm can be completed. To make the code better readable, each Yul dialect will have a default type which does not have to be explicitly mentioned. For eWasm this is u64, for EVM it is u256.

The Yul parser, analyzer and printer should all honour types and the analyzer should also check them (#2324).

It is to be discussed whether `bool` should be introduced as a type in the EVM dialect.

It should also be discussed how dialects deal with types for literals.

- [x] check that default types work both in parser and printer
- [x] perform type checking in analysis phase ( https://github.com/ethereum/solidity/issues/2324 )
- [ ] introduce types in optimizer phases (splitter, etc)
  - [ ] ControlFlowSimplifier::reduceSingleCaseSwitch: introduce typed dialect::equalityFunction and ::discardFunction and use. Same for all uses for makeDiscardCall in that file.
  - [x] ExpressionSplitter::outlineExpression needs to determine the type of the expression
  - [ ] ForLoopOutOfBody needs to select the proper negation function depending on the type.
  - [ ] test that FullInliner properly assigns types for new variables (for parameters and return parameters).
  - [ ] ssa transform needs to determine types for assignments.
  - [ ] VarDeclInitializer has to use `false` for bool variables.

### LINKED ISSUE #8154: [Yul] Remove the stack height checks from AsmAnalyzer
Direct access to the stack is not allowed any more, so we can remove most of the stack height checks.

Remove:

 - [x] AsmAnalysisInfo::stackHeightInfo
 - [x] AsmAnalyzer::m_stackHeight (some errors have to be checked differently, but we should also rephrase the error messages since they are all talking about the stack)

### PR DESCRIPTION
**Pull request by chriseth**
Fixes  #7070
Closes https://github.com/ethereum/solidity/issues/8154

### CODE REVIEW

File: libyul/Dialect.cpp [hunk 1]
```diff
@@ -19,10 +19,21 @@
...
 
 using namespace solidity::yul;
 using namespace std;
 
+bool Dialect::validTypeForLiteral(LiteralKind _kind, YulString, YulString _type) const
+{
+	if (_kind == LiteralKind::Boolean)
+		return _type == boolType;
+	else if (types.size() > 1)
+		return _type != boolType;
```
Comment (erak) [inline libyul/Dialect.cpp]: Would this return false for `2:bool`?

**Comment by leonardoalt** [comment]
Needs rebase

File: test/libyul/YulOptimizerTest.cpp [hunk 2]
```diff
@@ -349,7 +349,7 @@ TestCase::TestResult YulOptimizerTest::run(ostream& _stream, string const& _line
 	{
 		disambiguate();
 		ExpressionSplitter::run(*m_context, *m_ast);
-		WordSizeTransform::run(*m_dialect, *m_ast, *m_nameDispenser);
+		WordSizeTransform::run(*m_dialect, ""_yulstring, *m_ast, *m_nameDispenser);
```
Comment (leonardoalt) [inline test/libyul/YulOptimizerTest.cpp]: What happens with empty default type?

File: libyul/backends/wasm/WasmDialect.cpp [hunk 3]
```diff
@@ -56,34 +57,53 @@ WasmDialect::WasmDialect()
 	m_functions["i64.eq"_yulstring].returns.front() = "i32"_yulstring;
 	m_functions["i64.ne"_yulstring].returns.front() = "i32"_yulstring;
 
-	addFunction("i64.eqz", 1, 1);
+	addFunction("i32.eqz", "i32"_yulstring, 1, 1);
```
Comment (leonardoalt) [inline libyul/backends/wasm/WasmDialect.cpp]:
```suggestion
	addFunction("eqz", "i32"_yulstring, 1, 1);
```
Shouldn't it be like this?

File: libyul/AsmAnalysis.h [hunk 4]
```diff
@@ -112,6 +113,7 @@ class AsmAnalyzer
 	/// Variables that are active at the current point in assembly (as opposed to
 	/// "part of the scope but not yet declared")
 	std::set<Scope::Variable const*> m_activeVariables;
+	std::vector<YulString> m_typeOfCurrentExpression;
```
Comment (leonardoalt) [inline libyul/AsmAnalysis.h]: I don't understand fully how this vector is being used atm. Can you explain a bit?

File: test/libyul/YulOptimizerTest.cpp (hunk 2 above)
Comment (chriseth) [reply test/libyul/YulOptimizerTest.cpp]: This is the type that is used for the split variables. If it's empty then it's empty. Maybe I don't understand the question.

File: libyul/backends/wasm/WasmDialect.cpp (hunk 3 above)
Comment (chriseth) [reply libyul/backends/wasm/WasmDialect.cpp]: Why?

Comment (leonardoalt) [reply libyul/backends/wasm/WasmDialect.cpp]: I think this is related to me still not understanding what https://github.com/ethereum/solidity/pull/8161/files?file-filters%5B%5D=.h#diff-6311061a591e248cfc780cfd84bd3f22R61 does

Comment (chriseth) [reply libyul/backends/wasm/WasmDialect.cpp]: `i32.eqz` is the name of the wasm function. Some types, but not all, are part of the names of wasm functions. The parameter you linked is the type applied to all function parameters and return variables.

Comment (leonardoalt) [reply libyul/backends/wasm/WasmDialect.cpp]: Ah ok

**Review by leonardoalt — DISMISSED** [review]

**Comment by chriseth** [comment]
Yes, it's not finished either.

**Comment by chriseth** [comment]
Test for VerDeclInitializer:
```
{
    let a1
    let a2: bool
    let b1, b2: bool
    function f(a:u256, b:u256, c:bool) -> r:bool, t {
        let x1: bool, x2
    }
}
// ====
// dialect: evmTyped
// ----
```

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `03dfb0d9ca2317b335c86dc23627bf3c7fdb07f2`:
```
Error: Trailing whitespace found:
 test/libyul/yulSyntaxTests/invalid_type4.yul:2: switch 1 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `5124eea0b2c8ec5e82f82ec4fe59026b71a0034c`:
```
Error: Trailing whitespace found:
 test/libyul/yulSyntaxTests/invalid_type4.yul:2: switch 1 

```
Please check that your changes are working as intended.

File: libyul/Dialect.h [hunk 5]
```diff
@@ -52,15 +54,21 @@ struct Dialect: boost::noncopyable
...
 	YulString boolType;
-	std::set<YulString> types;
+	std::set<YulString> types = {{}};
 
 	/// @returns the builtin function of the given name or a nullptr if it is not a builtin function.
 	virtual BuiltinFunction const* builtin(YulString /*_name*/) const { return nullptr; }
 
-	virtual BuiltinFunction const* discardFunction() const { return nullptr; }
-	virtual BuiltinFunction const* equalityFunction() const { return nullptr; }
+	virtual BuiltinFunction const* discardFunction(YulString /* _type */) const { return nullptr; }
```
Comment (leonardoalt) [inline libyul/Dialect.h:62]: Why `/* _type */`?

File: test/libyul/Parser.cpp [hunk 6]
```diff
@@ -245,16 +245,6 @@ BOOST_AUTO_TEST_CASE(optional_types)
 	BOOST_CHECK(successParse("{ function f(a:u256) -> b {} }"));
 }
 
-BOOST_AUTO_TEST_CASE(invalid_types)
```
Comment (leonardoalt) [inline test/libyul/Parser.cpp:248]: Why was this removed?

Comment (leonardoalt) [reply test/libyul/Parser.cpp:248]: Ah looks like they were extracted.

File: test/libyul/yulSyntaxTests/type_check_cases_fail.yul [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+    switch 7:i32
```
Comment (leonardoalt) [inline test/libyul/yulSyntaxTests/type_check_cases_fail.yul:2]: Can you add a similar case but with `uint` vs `bool` in `evmTyped`?

File: test/libyul/yulSyntaxTests/for_loop_condition_fail.yul [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (leonardoalt) [inline test/libyul/yulSyntaxTests/for_loop_condition_fail.yul:1]: Would it work to add a similar test but for `ewasm`?

**Review by leonardoalt — COMMENTED** [review]
Looks good!
A few more tests (as requested) would be nice (if they make sense).

File: libyul/Dialect.h (hunk 5 above)
Comment (chriseth) [reply libyul/Dialect.h:62]: It is unused in this base implementation, but some dialects have a specific discard function per type.

File: test/libyul/Parser.cpp (hunk 6 above)
Comment (chriseth) [reply test/libyul/Parser.cpp:248]: I turned it into a YulSyntaxTest, it was easier to do there.

**Comment by chriseth** [comment]
@leonardoalt please check again.

**Comment by chriseth** [comment]
(just added the two tests)

**Review by leonardoalt — APPROVED** [review]
